### DISCUSSION TITLE
80186 implementation of timer2 terminal count latch/handler for dma control

### DISCUSSION THREAD

**Discussion by hjanetzek** [mamedev/discussions#99] [General]
Hello, this is my first deeper dive into MAMEs codebase. I wonder how much work it would be to add support for TIMER_DRQ as described in section 4.6 (page 30) here. Hints or suggestions how to approach this and what to take care of would be very welcome 
http://bitsavers.informatik.uni-stuttgart.de/components/intel/80186/210973-001_AP-186_Introduction_to_the_80186_Microprocessor_Mar83.pdf

**Comment by hjanetzek** [comment]
Looking at https://github.com/mamedev/mame/blob/master/src/devices/cpu/i86/i186.cpp#L1491
There appear to be a few more parts missing
```
	if (diff & 0x6811)
		LOGMASKED(LOG_DMA, "%05X:ERROR! - unsupported DMA mode %04X\n", m_pc, new_control);
        // 0110 1000 0001 0001 Unsupported?
        // M/~IO Destination pointer is in M/IO space
        //  Destination decrement
        //   Destination increment
        //    M/~IO Souce pointer is in M/IO space
        //     Source decrement
        //      Source increment
        //              TDRQ Enable DMA requests from timer !!!!
```
My ROM needs these to work
```
        // DMA0
        // 1111 0111 0001 0100
        // 1111 0111 0001 0010
        //              Enable DMA Requests from timer2
```

**Comment by cracyc** [comment]
That's not a fatal error and it's wrong too.  All of those bits are implemented including timer drq, https://github.com/mamedev/mame/blob/master/src/devices/cpu/i86/i186.cpp#L1261.  I don't think timer drqs are well tested though so it might not work correctly.

**Comment by hjanetzek** [comment]
Ah ok could this case in drq_callback be wrong? 
[:maincpu] F989F:ERROR! - drq0 with dma channel stopped

```
TIMER_CALLBACK_MEMBER(i80186_cpu_device::timer_elapsed)
...
          if ((m_dma[0].control & (TIMER_DRQ | ST_STOP)) == TIMER_DRQ) {
            drq_callback(0);
          }
```
^^^ ST_STOP must not be set and below it must be set
 
```
void i80186_cpu_device::drq_callback(int which)
...
       if (!(dma->control & ST_STOP))
	{
             LOGMASKED(LOG_DMA, "%05X:ERROR! - drq%d with dma channel stopped\n", m_pc, which);      
	     return;
	}
```

**Reply by cracyc** [reply]
I think timer_elapsed is wrong and ST_STOP must be set.

**Reply by hjanetzek** [reply]
This is certainly a good starting point - with `if ((m_dma[0].control & (TIMER_DRQ | ST_STOP)) ==  (TIMER_DRQ | ST_STOP))` the display sometimes renders correctly, still updates are flickering and garbled. Before it didn't render anything without adding an external timer to trigger drq0. I'll try a bit further

**Reply by hjanetzek** [reply]
It could be the case that this is the only change needed to make TDRQ work and that the issue I'm seeing with the garbled display is due to the program reinitializes from an exception unrelated to DMA while the transfer/timer still runs

**Reply by hjanetzek** [reply]
yep, this was indeed the bug

**Comment by hjanetzek** [comment]
Hey @cracyc should I make a PR for this change or do you need some device to verify it? There appears also to be a bug when timer2 is used for pre-scaling where timer0/1 are still restarted with `t->int_timer->adjust()` resulting in a way too fast timer. This makes it work for me
```
void i80186_cpu_device::restart_timer(int which)
{
	timer_state *t = &m_timer[which];
	int count = (t->control & 0x1000) ? t->maxB : t->maxA;
	if (!(t->control & 4)) {
        // Only run timer0/1 when not incremented by timer2 pre-scaler
        if (!((t->control & 0x800c) == 0x8008)) {
            t->int_timer->adjust((attotime::from_hz(clock() / 8) * (count ? count : 0x10000)),which);
        }
    }    
}
```

**Comment by cracyc** [comment]
I'd say go ahead and make a pr.  Your thing is the only thing I'm aware of that uses those features.
